src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestExecutor.cs

```diff
@@ -47,6 +48,20 @@ internal MSTestExecutor(CancellationToken cancellationToken)
     /// </summary>
     public TestExecutionManager TestExecutionManager { get; protected set; }
 
+#pragma warning disable CA2255 // The 'ModuleInitializer' attribute should not be used in libraries
+    [ModuleInitializer]
+#pragma warning restore CA2255 // The 'ModuleInitializer' attribute should not be used in libraries
+    internal static void EnsureAdapterAndFrameworkVersions()
+    {
+        string? adapterVersion = typeof(MSTestExecutor).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+        string? frameworkVersion = typeof(TestMethodAttribute).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+        if (adapterVersion is not null && frameworkVersion is not null
+            && adapterVersion != frameworkVersion)
+        {
+            throw new InvalidOperationException($"MSTest.TestAdapter and MSTest.TestFramework must have the same version. Found MSTest.TestAdapter version '{adapterVersion}' and MSTest.TestFramework version {frameworkVersion}.");
```

Comment:
Message should be localized. I would do a nit and add quotes around package names.